src/collection.jl

```diff
@@ -0,0 +1,100 @@
+# helper function to match extension
+# this will be removed when there is a compat entry of endswith with regex compatibility
+function match_extension(str::String, ext::Regex)
+    r = deepcopy(ext)
+    if !endswith(r.pattern, "\$")
+        r = Regex(r.pattern * "\$", r.compile_options, r.match_options)
+    end
+    return occursin(r, str)
+end
+
+match_extension(str::String, ext::String) = endswith(str, ext)
+
+#------------------------------------------------------------------------------------------------
+@doc raw"""
+    fitscollection(dir; recursive=true, abspath=true, keepext=true, ext=r"fits(\.tar\.gz)?", exclude=nothing, exclude_dir=nothing, exclude_key = ("", "HISTORY"))
+
+Walk through `dir` collecting FITS files, scanning their headers, and culminating into a `DataFrame` that can be used with the generators for iterating over many files and processing them. If `recursive` is false, no subdirectories will be walked through.
+
+The table returned will contain the path to the file, the name of the file, and index of the corresponding hdu, and each FITS header column and value. If two FITS files have distinct columns, they will both appear in the table with `missing` in the appropriate rows.
+
+If `abspath` is true, the path in the table will be absolute. If `keepext` is true, the name in the table will include the file extension, given by `ext`. `ext` will be used with `endswith` to filter for fits files compatible with `FITSIO.FITS`. `exclude` is a pattern that can be used with `occursin` to exclude certain filenames. For example, to exclude any files containing "sky",
+```julia
+fitscollection(...; exclude="sky")
+```
+to exclude exact filenames, [regex strings](https://docs.julialang.org/en/v1/manual/strings/#Regular-Expressions-1) will prove powerful
+```julia
+fitscollection(...; exclude=r"^tek001\d")
+```
+finally, using external tools like [Glob.jl](https://github.com/vtjnash/Glob.jl) allows further customization
+```julia
+using Glob
+fitscollection(...; exclude=fn"tek001*.fits") # same as regex match above
+```
+Similarly, `exclude_dir` allows excluding entire folders using pattern matching (e.g. skipping a backup folder `exclude_dir="backup"`).
+`exclude_key` allows excluding certain entries in the header unit of `ImageHDU` in FITS files (e.g. skipping `"HISTORY"` and `""` `exclude_key = ("HISTORY", "")`).
+
+For more information about the file matching and path deconstruction, see the extended help (`??fitscollection`)
+# Extended Help
+
+## Parts of a path
+
+Here is an example of a file path and how it would be parsed
+
+```
+     root         dir   base   ext
+[---------------][----][-----][----]
+\Users\miles\data\test\tek0001.fits
+
+     root          dir     base    ext
+[---------------][--------][-----][----]
+\Users\miles\data\test\sci\tek0001.fits
+```
```

Comment:
I think this can be cleaned up a little-

```julia
Let's look at some file paths starting from `"/data"`. Here are examples of how they would be parsed

\```
 root  dir   base   ext
[----][---][------][---]
/data/test/tek0001.fits

 root    dir     base   ext
[----][-------][------][---]
/data/test/sci/tek0001.fits
\```
```